File: samples/java/SupportChat/Server/Configuration/src/main/java/systems/zlink/samples/supportchat/server/configuration/SupportChatReadinessReporter.java
package systems.zlink.samples.supportchat.server.configuration;

import org.springframework.boot.ApplicationArguments;
import org.springframework.boot.ApplicationRunner;

import systems.zlink.framework.errors.ZLinkConfigurationException;
import systems.zlink.framework.monitoring.ZLinkPeerState;
import systems.zlink.framework.monitoring.ZLinkRouteMeshRuntime;

import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;

public final class SupportChatReadinessReporter implements ApplicationRunner, AutoCloseable {
    private final String nodeId;
    private final ZLinkRouteMeshRuntime meshes;
    private final ScheduledExecutorService reporter =
            Executors.newSingleThreadScheduledExecutor(
                    runnable -> {
                        Thread thread = new Thread(runnable, "supportchat-readiness");
                        thread.setDaemon(true);
                        return thread;
                    });

    public SupportChatReadinessReporter(String nodeId, ZLinkRouteMeshRuntime meshes) {
        this.nodeId = nodeId;
        this.meshes = meshes;
    }

    @Override
    public void run(ApplicationArguments arguments) {
        reporter.scheduleWithFixedDelay(this::report, 0, 100, TimeUnit.MILLISECONDS);
    }

    private void report() {
        try {
            boolean supportRouteReady =
                    meshes.snapshot(SampleNames.SupportActorMesh).peers().stream()
                            .anyMatch(
                                    peer ->
                                            peer.nodeRid().equals(SampleNames.SupportNodeRoutingId)
                                                    && peer.state() == ZLinkPeerState.READY);
            if (!supportRouteReady) {
                return;
            }
            System.out.printf(
                    "supportchat-ready kind=spot-route node=%s mesh=%s%n",
                    nodeId, SampleNames.SupportActorMesh);
            reporter.shutdown();
        } catch (IllegalStateException | ZLinkConfigurationException ignored) {
            // The public runtime view is not available until Framework startup completes.
        }
    }

    @Override
    public void close() {
        reporter.shutdownNow();
    }
}

File: samples/kotlin/SupportChat/Server/Configuration/src/main/kotlin/systems/zlink/samples/kotlin/supportchat/server/configuration/SupportChatReadinessReporter.kt
package systems.zlink.samples.kotlin.supportchat.server.configuration

import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit
import org.slf4j.LoggerFactory
import org.springframework.boot.ApplicationArguments
import org.springframework.boot.ApplicationRunner
import systems.zlink.framework.errors.ZLinkConfigurationException
import systems.zlink.framework.monitoring.ZLinkRouteMeshRuntime

class SupportChatReadinessReporter(
    private val nodeId: String,
    private val meshes: ZLinkRouteMeshRuntime,
) : ApplicationRunner, AutoCloseable {
    private val reporter =
        Executors.newSingleThreadScheduledExecutor { runnable ->
            Thread(runnable, "supportchat-readiness").apply { isDaemon = true }
        }

    override fun run(args: ApplicationArguments) {
        reporter.scheduleWithFixedDelay(::report, 0, 100, TimeUnit.MILLISECONDS)
    }

    private fun report() {
        try {
            val mesh = meshes.snapshot(SampleNames.SupportSpotDiscovery)
            if (!mesh.isReady || mesh.readyPeerCount == 0) return

            logger.info(
                "supportchat-ready kind=spot-route node={} mesh={}",
                nodeId,
                SampleNames.SupportSpotDiscovery,
            )
            reporter.shutdown()
        } catch (_: IllegalStateException) {
            // The public runtime view is not available until Framework startup completes.
        } catch (_: ZLinkConfigurationException) {
            // The public runtime view is not available until Framework startup completes.
        }
    }

    override fun close() {
        reporter.shutdownNow()
    }

    private companion object {
        private val logger = LoggerFactory.getLogger(SupportChatReadinessReporter::class.java)
    }
}
